### PR TITLE
error messages: report a list of missing overrides

### PR DESCRIPTION
**Pull request by smoothdeveloper**
This is going to be for #4982, right now, it is just a first pass at looking at the code and cleaning a bit without any behaviour change.

I intend to look at `noimpl` which is the location we report the error.

Instead I plan to make it an accumulator which will raise the error (with the expected message) if more than 10 missing overrides have been encountered, otherwise it will append, and then it will be checked before returning to construct the expected message if needed and raise it.

Please let me know if you see any issue with that plan.

Status:
* [x] main logic implementation
* [x] reorganize error message (and fix their names)
* [x] add exhaustive tests for each variant of the message
* [x] make it green

### CODE REVIEW

File: src/fsharp/MethodOverrides.fs [hunk 1]
```diff
@@ -255,23 +255,39 @@ module DispatchSlotChecking =
...
+            if showMissingMethodsAndRaiseErrors then
+                errorR exn
+
+        let rec listSameLength l1 l2 =
+          let l1Empty = List.isEmpty l1
+          let l2Empty = List.isEmpty l2
+          if l1Empty && l2Empty then true
+          elif l1Empty || l2Empty then false
+          else
+            listSameLength l1.Tail l2.Tail
```
Comment (vasily-kirichenko) [inline src/fsharp/MethodOverrides.fs]:
```fsharp

let rec listSameLength l1 l2 = 
    match l1, l2 with
    | [], [] -> true
    | [], _ | _, [] -> false
    | _ :: t1, _ :: t2 -> listSameLength t1 t2
```
?

File: src/fsharp/MethodOverrides.fs [hunk 2]
```diff
@@ -308,28 +324,28 @@ module DispatchSlotChecking =
                                 // Error will be reported below in CheckOverridesAreAllUsedOnce 
                                 ()
 
-                            elif argTys.Length <> vargtys.Length then 
-                                fail(Error(FSComp.SR.typrelMemberDoesNotHaveCorrectNumberOfArguments(FormatOverride denv overrideBy, FormatMethInfoSig g amap m denv dispatchSlot), overrideBy.Range))
+                            elif not (listSameLength argTys vargtys) then 
+                                fail(Error(FSComp.SR.typrelMemberDoesNotHaveCorrectNumberOfArguments(FormatOverride denv overrideBy, formatMethodInfoSignature dispatchSlot), overrideBy.Range))
                             elif mtps.Length <> fvmtps.Length then
```
Comment (vasily-kirichenko) [inline src/fsharp/MethodOverrides.fs]: If `mtps` is a list, use `listSameLength` here as well.

Comment (vasily-kirichenko) [reply src/fsharp/MethodOverrides.fs]: I think there is already an issue/suggestion. It would be easy to implement, the hard part is to decide how to represent it on the tooltip.

Comment (smoothdeveloper) [reply src/fsharp/MethodOverrides.fs]: #3110 revolves around similar topic, in this case it would display:

image
> Typars = Typar list

**Comment by smoothdeveloper** [comment]
I've added the main logic to `CheckDispatchSlotsAreImplemented`, few decisions to make:
* do we want to keep the same compiler error number among the variants of the message?
* do we want to reduce the variants of message regarding the one with suggestion related to interface? I believe this should be handled differently by appending another message rather than having a couple of redundant messages with just the end being different (see changes in FSComp.txt)

I'll be working on few code samples to add to the test suite, is `fsharpqa\Source\ErrorMessages\MissingOverrides` a good location for it?

If you have evil code idea to check I'm not breaking anything or that we have all cases cornered, please share.

Sample output for now:
```fsharp
Microsoft (R) F# Interactive version 10.1.0 for F# 4.1
Copyright (c) Microsoft Corporation. All Rights Reserved.

For help type #help;;

> open System
- open System.IO;;
>
- type IFoo1 =
-   abstract member Foo : t:Type * r:TextReader -> obj
-   abstract member Bar<'t> : TextReader -> 't
-   abstract member Baz<'t> : TextReader -> 't
-
- type Foo1() =
-   interface IFoo1 with
-     member x.Foo(t, reader) = obj()
- ;;

    interface IFoo1 with
  ------------^^^^^

stdin(10,13): error FS0366: No implementation was given for those members:
        'abstract member IFoo1.Bar : TextReader -> 't'
        'abstract member IFoo1.Baz : TextReader -> 't'
 Note that all interface members must be implemented and listed under an appropriate 'interface' declaration, e.g. 'interface ... with member ...'.

> type IFoo2 =
-   abstract member Foo : t:Type * r:TextReader -> obj
-   abstract member Bar<'t> : TextReader -> 't
-   abstract member Baz<'t> : TextReader -> 't
-   abstract member Baz1<'t> : TextReader -> 't
-   abstract member Baz2<'t> : TextReader -> 't
-   abstract member Baz3<'t> : TextReader -> 't
-   abstract member Baz4<'t> : TextReader -> 't
-   abstract member Baz5<'t> : TextReader -> 't
-   abstract member Baz6<'t> : TextReader -> 't
-   abstract member Baz7<'t> : TextReader -> 't
-   abstract member Baz8<'t> : TextReader -> 't
-   abstract member Baz9<'t> : TextReader -> 't
-
- type Foo2() =
-   interface IFoo2 with
-     member x.Foo(t, reader) = obj()
-     ;;

    interface IFoo2 with
  ------------^^^^^

stdin(28,13): error FS0366: No implementation was given for those members (some results omitted):
        'abstract member IFoo2.Bar : TextReader -> 't'
        'abstract member IFoo2.Baz : TextReader -> 't'
        'abstract member IFoo2.Baz1 : TextReader -> 't'
        'abstract member IFoo2.Baz2 : TextReader -> 't'
        'abstract member IFoo2.Baz3 : TextReader -> 't'
        'abstract member IFoo2.Baz4 : TextReader -> 't'
        'abstract member IFoo2.Baz5 : TextReader -> 't'
        'abstract member IFoo2.Baz6 : TextReader -> 't'
        'abstract member IFoo2.Baz7 : TextReader -> 't'
        'abstract member IFoo2.Baz8 : TextReader -> 't'
 Note that all interface members must be implemented and listed under an appropriate 'interface' declaration, e.g. 'interface ... with member ...'.
>
```

**Comment by smoothdeveloper** [comment]
I'm not understanding why CI part 3 & 4 fail with this error:

```
19:59:20 D:\j\w\release_ci_pa---c6929ad7\src\fsharp\MethodOverrides.fs(376,48): error FS0039: The field, constructor or member 'typrelNoImplementationSeveralGivenWithSuggestion' is not defined. Maybe you want one of the following:   typrelNoImplementationGivenWithSuggestion   typrelNoImplementationGiven   typrelCannotResolveImplicitGenericInstantiation   typrelOverrideImplementsMoreThenOneSlot   typrelNamedArgumentHasBeenAssignedMoreThenOnce [D:\j\w\release_ci_pa---c6929ad7\src\buildfromsource\FSharp.Compiler.Private\FSharp.Compiler.Private.fsproj]
19:59:20 D:\j\w\release_ci_pa---c6929ad7\src\fsharp\MethodOverrides.fs(377,49): error FS0039: The field, constructor or member 'typrelNoImplementationSeveralGiven' is not defined. Maybe you want one of the following:   typrelNoImplementationGiven   typrelNoImplementationGivenWithSuggestion   typrelCannotResolveAmbiguityInDelegate   typrelOverrideImplementsMoreThenOneSlot   typrelCannotResolveImplicitGenericInstantiation [D:\j\w\release_ci_pa---c6929ad7\src\buildfromsource\FSharp.Compiler.Private\FSharp.Compiler.Private.fsproj]
19:59:20 D:\j\w\release_ci_pa---c6929ad7\src\fsharp\MethodOverrides.fs(378,47): error FS0039: The field, constructor or member 'typrelNoImplementationSeveralTruncatedGivenWithSuggestion' is not defined. Maybe you want one of the following:   typrelNoImplementationGivenWithSuggestion   typrelCannotResolveImplicitGenericInstantiation   typrelTypeImplementsIComparableShouldOverrideObjectEquals   typrelMemberDoesNotHaveCorrectKindsOfGenericParameters   typrelTypeImplementsIComparableDefaultObjectEqualsProvided [D:\j\w\release_ci_pa---c6929ad7\src\buildfromsource\FSharp.Compiler.Private\FSharp.Compiler.Private.fsproj]
19:59:20 D:\j\w\release_ci_pa---c6929ad7\src\fsharp\MethodOverrides.fs(379,48): error FS0039: The field, constructor or member 'typrelNoImplementationSeveralTruncatedGiven' is not defined. Maybe you want one of the following:   typrelNoImplementationGivenWithSuggestion   typrelNoImplementationGiven   typrelSigImplNotCompatibleConstraintsDiffer   typrelSigImplNotCompatibleConstraintsDifferRemove   typrelCannotResolveImplicitGenericInstantiation [D:\j\w\release_ci_pa---c6929ad7\src\buildfromsource\FSharp.Compiler.Private\FSharp.Compiler.Private.fsproj]
```

those are defined in https://github.com/Microsoft/visualfsharp/pull/4983/files#diff-0e57ebf050db850b5efdd7e2cb13e725 and it compiles in part 1 & 2.

**Comment by dsyme** [comment]
> I'm not understanding why CI part 3 & 4 fail with this error:


You have to update the FSComp files at least.  See DEVGUIDE.md
```
    .\build net40
    copy /y src\fsharp\FSharp.Compiler.Private\obj\release\net40\FSComp.* src\buildfromsource\FSharp.Compiler.Private\
```

File: src/fsharp/MethodOverrides.fs [hunk 3]
```diff
@@ -255,23 +255,40 @@ module DispatchSlotChecking =
...
-        let res = ref true
-        let fail exn = (res := false ; if showMissingMethodsAndRaiseErrors then errorR exn)
         
+        let mutable res = true
+        let fail exn =
+            res <- false
+            if showMissingMethodsAndRaiseErrors then
+                errorR exn
+
+        let rec listSameLength l1 l2 =
```
Comment (dsyme) [inline src/fsharp/MethodOverrides.fs]: Just use `List.length l1 = List.length l2` or `l1.Length = l2.Length`. No need for a new function

File: src/fsharp/MethodOverrides.fs [hunk 4]
```diff
@@ -255,23 +255,40 @@ module DispatchSlotChecking =
...
+            if showMissingMethodsAndRaiseErrors then
+                errorR exn
+
+        let rec listSameLength l1 l2 = 
+            match l1, l2 with
+            | [], [] -> true
+            | [], _ | _, [] -> false
+            | _ :: t1, _ :: t2 -> listSameLength t1 t2
+
+        let inline formatMethodInfoSignature dispatchSlot =
```
Comment (dsyme) [inline src/fsharp/MethodOverrides.fs]: Don't use `inline`.  Indeed don't define this local function, it makes code harder to understand to be honest.

File: src/fsharp/MethodOverrides.fs [hunk 5]
```diff
@@ -308,28 +322,65 @@ module DispatchSlotChecking =
                                 // Error will be reported below in CheckOverridesAreAllUsedOnce 
                                 ()
 
-                            elif argTys.Length <> vargtys.Length then
```
Comment (dsyme) [inline src/fsharp/MethodOverrides.fs]: just revert these changes I think

File: src/fsharp/MethodOverrides.fs [hunk 6]
```diff
@@ -308,28 +322,65 @@ module DispatchSlotChecking =
...
-                                fail(Error(FSComp.SR.typrelMemberDoesNotHaveCorrectKindsOfGenericParameters(FormatOverride denv overrideBy, FormatMethInfoSig g amap m denv dispatchSlot), overrideBy.Range))
+                                fail(Error(FSComp.SR.typrelMemberDoesNotHaveCorrectKindsOfGenericParameters(FormatOverride denv overrideBy, formatMethodInfoSignature dispatchSlot), overrideBy.Range))
                             else 
-                                fail(Error(FSComp.SR.typrelMemberCannotImplement(FormatOverride denv overrideBy, NicePrint.stringOfMethInfo amap m denv dispatchSlot, FormatMethInfoSig g amap m denv dispatchSlot), overrideBy.Range))
+                                fail(Error(FSComp.SR.typrelMemberCannotImplement(FormatOverride denv overrideBy, NicePrint.stringOfMethInfo amap m denv dispatchSlot, formatMethodInfoSignature dispatchSlot), overrideBy.Range))
                         | overrideBy :: _ -> 
-                            errorR(Error(FSComp.SR.typrelOverloadNotFound(FormatMethInfoSig g amap m denv dispatchSlot, FormatMethInfoSig g amap m denv dispatchSlot),overrideBy.Range))
```
Comment (dsyme) [inline src/fsharp/MethodOverrides.fs]: revert these ones too

File: build.cmd [hunk 7]
```diff
@@ -79,6 +79,8 @@ set TEST_END_2_END=0
 set INCLUDE_TEST_SPEC_NUNIT=
 set INCLUDE_TEST_TAGS=
 
+set COPY_FSCOMP_RESOURCE_FOR_BUILD_FROM_SOURCES=0
```
Comment (smoothdeveloper) [inline build.cmd]: @KevinRansom can you check if you like this and if it is OK? Thanks!

File: src/buildfromsource/FSharp.Compiler.Private/FSComp.resx [hunk 8]
```diff
@@ -4343,6 +4355,6 @@
     <value>A type annotated with IsReadOnly must also be a struct. Consider adding the [&lt;Struct&gt;] attribute to the type.</value>
   </data>
   <data name="chkStructsMayNotReturnAddressesOfContents" xml:space="preserve">
-    <value>Struct members cannot return 'this' or fields by reference</value>
+    <value>Struct members cannot return the address of fields of the struct by reference</value>
```
Comment (smoothdeveloper) [inline src/buildfromsource/FSharp.Compiler.Private/FSComp.resx]: not sure why this line is changed, might be due to rebase or inconsistent file (we shouldn't checkin generated files).

File: build.cmd [hunk 9]
```diff
@@ -470,6 +474,7 @@ if /i "%BUILD_CORECLR%" == "1" (
 
 if /i "%BUILD_FROMSOURCE%" == "1" (
     set NEEDS_DOTNET_CLI_TOOLS=1
+    set COPY_FSCOMP_RESOURCE_FOR_BUILD_FROM_SOURCES=1
```
Comment (dsyme) [inline build.cmd]: No - do this for the **non-build-from-source** cases.  The copy is intended to prepare the generated FSComp files for checkin when you do something like `build net40`

File: src/buildfromsource/FSharp.Compiler.Private/FSComp.resx (hunk 8 above)
Comment (dsyme) [reply src/buildfromsource/FSharp.Compiler.Private/FSComp.resx]: Don't worry about it, some file didn't get updated when it should have been, this is now updating it

Comment (smoothdeveloper) [reply src/buildfromsource/FSharp.Compiler.Private/FSComp.resx]: Ok, but then can we make a plan to remove the files that are generated after an initial `build` from the repository? 

I think it would help to know what is canonical file, remove burden in code reviews and improve other maintainability aspects.

File: build.cmd (hunk 9 above)
Comment (smoothdeveloper) [reply build.cmd]: I'm not sure about this particular line, but this change to build.cmd actually seems to makes the CI pass the "build from source" test without manually having to take care of those files, I'm probably missing something but ideally that is what should happen without any intervention from contributors.

Unless for some reason, we'd have different messages going there.

Should I remove this line and check that it is still passing the build from source test?

Comment (smoothdeveloper) [reply build.cmd]: I've removed that line from #5119, I'll track there that whole change for build-from-source fscomp files.

File: build.cmd (hunk 7 above)
Comment (KevinRansom) [reply build.cmd]: @smoothdeveloper this is something that we need to fix in buildfromsource, although, this seems like a pretty decent work around.

The reason the copy is needed is that the buildtask used for the RedHat build doesn't yet or at least, didn't the last time we looked support the resource id generation.

@brettfo are we in a position to update this yet?

File: build.cmd (hunk 9 above)
Comment (KevinRansom) [reply build.cmd]: @smoothdeveloper Don, has this correct, when building net40, copy over the files.  until we can fix the  redhat build.

File: build.cmd (hunk 7 above)
Comment (smoothdeveloper) [reply build.cmd]: @KevinRansom thanks, yes if in the future the build from source can prepare the package by doing the copy for us from the canonical files at their current place, it would remove the duplicates and some noise in each of the changes to FSComp.

The other pain area is the localization files, it is great that they are open source and fully integrated in the delivery process, but we should avoid the need to check them in this repository, or at least have them automatically updated in our CI, there can be another localization CI which doesn't make our PR here red and is handled by Redmond for the actual release pipeline.

Hope that makes sense, thanks all for reviewing.

**Comment by KevinRansom** [comment]
@smoothdeveloper,

thanks for this

Kevin
